feat: add Trogon.Ecto package

yordis opened the pull request.

yordis committed:
feat: add OnePiece.Ecto package

yordis committed:
feat(trogon_ecto): introduce Trogon.Ecto.ValueObject

- Encode the DDD value-object intent in the module name so call sites read as constructors of immutable values rather than as generic schema helpers.
- Reject foreign structs at every entry point so wrong-type values fail loudly with a descriptive error instead of being silently coerced through Map.from_struct.
- Trust own-type structs as already validated (put_embed semantics) so VOs stay immutable end-to-end and we never re-run validation on values that, by construction, were valid.

Signed-off-by: Yordis Prieto <yordis.prieto@gmail.com>

yordis committed:
refactor(trogon_ecto): address review feedback on ValueObject

Surface a descriptive ArgumentError on foreign struct input and accept
own-type structs as immutable so callers get predictable factory
semantics instead of an opaque FunctionClauseError.

Signed-off-by: Yordis Prieto <yordis.prieto@gmail.com>

yordis commented on apps/trogon_ecto/lib/trogon/ecto/value_object.ex:
```diff
@@ -0,0 +1,405 @@
+    struct(struct_module)
+    |> struct_module.changeset(attrs)
+  end
+end
```
Intentionally, will be move to it later

yordis committed:
fix(trogon_ecto): enforce required polymorphic_embeds_many

PolymorphicEmbed's required option does not treat an empty list as
missing, so enforced *_many fields silently accepted no entries.

Signed-off-by: Yordis Prieto <yordis.prieto@gmail.com>

yordis committed:
fix(trogon_ecto): override embed_as to dump nested value objects

Ecto.Type defaults embed_as to :self, so nested value objects used as
custom-type fields were stored as unconverted structs, breaking JSON
serialization on persistence.

Signed-off-by: Yordis Prieto <yordis.prieto@gmail.com>

enilsen16 commented on apps/trogon_ecto/lib/trogon/ecto/value_object.ex:
```diff
@@ -0,0 +1,428 @@
+      def load(value) when is_struct(value, __MODULE__), do: {:ok, value}
+      def load(%_other{}), do: :error
+
+      def load(data) when is_map(data) do
```
`load/1` can raise instead of returning `:error` for a malformed map, which is inconsistent with its sibling clauses. `load(%_other{})` and `load(_)` both return `:error`, and the tests assert `:error` for bad input (`load(1)`), but this `is_map` clause calls `Ecto.embedded_load/3`, which **raises** rather than returning `:error`. Example: a polymorphic-embed map missing a resolvable `__type__` discriminator hits PolymorphicEmbed's `raise_cannot_infer_type_from_data` (a `RuntimeError`); a plain field with an unloadable value raises too.

A direct caller relying on the documented `{:ok, _} | :error` contract gets an exception instead. (In the `Repo` read path this is moot — Ecto's `load!` raises on `:error` anyway — so it only bites direct/manual `load/1` calls and the contract the surrounding clauses imply.) Wrapping the body in a `try/rescue` that maps the raise to `:error` would make this clause consistent with the others.

enilsen16 commented on apps/trogon_ecto/mix.exs:
```diff
@@ -0,0 +1,118 @@
+
+  defp deps do
+    [
+      {:ecto, "~> 3.6"},
```
`{:ecto, "~> 3.6"}` understates the real floor. `polymorphic_embed_field?/2` and `polymorphic_embed_many?/2` match the parameterized-type shape `{:parameterized, {PolymorphicEmbed, _}}`, which Ecto only introduced in **3.12** (pre-3.12 used the 3-tuple `{:parameterized, PolymorphicEmbed, _}`). On an older Ecto those clauses silently fall through to `_ -> false`, so every polymorphic field is treated as a plain field — no error, just dropped casting and required-validation.

The lockfile resolves 3.13.5 and `polymorphic_embed ~> 5.0` transitively forces `ecto ~> 3.12`, so this is latent today. Still, the declared constraint should be honest: `~> 3.12`.

enilsen16 commented on apps/trogon_ecto/lib/trogon/ecto/value_object.ex:
```diff
@@ -0,0 +1,428 @@
+
+  def new(struct_module, %other{}) when is_atom(struct_module) do
+    raise ArgumentError,
+          "expected attrs to be a map or %#{inspect(struct_module)}{}, got %#{inspect(other)}{}"
```
This error string is hand-duplicated three times (here, plus lines 307 and 347). Any wording change has to stay byte-identical across all three — and the doctests/tests that assert on it. Consider a small private helper, e.g. `raise_unexpected_attrs(struct_module, other)`, called from `new/2`, `new!/2`, and `changeset/2`.

yordis committed:
refactor(trogon_ecto): address review feedback on ValueObject

Signed-off-by: Yordis Prieto <yordis.prieto@gmail.com>

enilsen16 commented on apps/trogon_ecto/lib/trogon/ecto/value_object.ex:
```diff
@@ -0,0 +1,432 @@
+    PolymorphicEmbed.cast_polymorphic_embed(changeset, field, required: struct_module.__enforced_keys__?(field))
+  end
+
+  defp get_polymorphic_embeds(struct_module) do
```
`get_polymorphic_embeds/1` (plus `polymorphic_embed_field?/2` and `polymorphic_embed_many?/2`) re-scans every field through `__schema__(:type, field)` on each `changeset/2` call — i.e. on every `new/0`. The set of polymorphic-embed fields (and which are `_many`) is fixed at compile time.

`__before_compile__` already precomputes the analogous static fact (the enforced-key set) — so this could follow the same pattern: emit `__polymorphic_embeds__/0` / `__polymorphic_embeds_many__/0` there and have `changeset/2` look them up, turning per-construction O(fields) work into an O(1) lookup.

Optional / consistency rather than a hot-path concern — VOs are typically small. Everything else here checks out; nice package.

enilsen16 reviewed (approved):
No correctness issues — verified the suspect paths against Ecto 3.12 / polymorphic_embed v5 source (apply_action(:new), the polymorphic `_many` required gap-filler, the parameterized tuple shape, and embeds_many required all check out). One optional consistency suggestion inline re: precomputing the polymorphic-embed field set in `__before_compile__`. Nice package.

yordis committed:
Cache polymorphic embeds at compile time

Resolve polymorphic embed fields during compilation rather than
inspecting the schema on every changeset call, avoiding repeated
runtime introspection.

Signed-off-by: Yordis Prieto <yordis.prieto@gmail.com>

yordis committed:
refactor(trogon_ecto): precompute changeset metadata at compile time

Signed-off-by: Yordis Prieto <yordis.prieto@gmail.com>
